domain/upgrade/service/upgrade_watcher.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/worker/v3/catacomb"
+
+	"github.com/juju/juju/core/changestream"
+	"github.com/juju/juju/core/watcher"
+)
+
+type upgradeWatcher struct {
+	ctx      context.Context
+	catacomb catacomb.Catacomb
+
+	st          State
+	wf          WatcherFactory
+	upgradeUUID string
+
+	out chan struct{}
+}
+
+// NewUpgradeWatcher creates a watcher which notifies when controller node
+// which owns the upgrade - complete the process.
+func NewUpgradeWatcher(ctx context.Context, st State, wf WatcherFactory, upgradeUUID string) (watcher.NotifyWatcher, error) {
+	w := &upgradeWatcher{
+		ctx:         ctx,
+		st:          st,
+		wf:          wf,
+		upgradeUUID: upgradeUUID,
+		out:         make(chan struct{}),
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return w, nil
+}
+
+func (w *upgradeWatcher) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+func (w *upgradeWatcher) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *upgradeWatcher) Changes() <-chan struct{} {
+	return w.out
+}
+
+func (w *upgradeWatcher) loop() error {
+	defer close(w.out)
+
+	uuidWatcher, err := w.wf.NewUUIDsWatcher("upgrade_info", changestream.Create|changestream.Update)
```

Comment:
Following on from what we discussed yesterday, we are not going to use this custom watcher.

We will simply have a method on the service, `WatchUpgradeInfo` and it will call `NewValueWatcher("upgrade_info", uuid, changestream.Create|changestream.Update)`

`NewUUIDsWatcher` is a namespace watcher, which watches the whole table.